### PR TITLE
Feature filter2manager

### PR DESCRIPTION
**Pull request by toros11**
see #364

In addition added unit/integration tests and support for adding items of mode type
through vnctl.

### CODE REVIEW

File: vnet/db/migrations/0006_filters.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+      Integer :interface_id, :index => true
+      
+      FalseClass :ingress_passthrough, :null=> false
+      FalseClass :egress_passthrough, :null=>false
+
+      DateTime :created_at, :null =>false
+      DateTime :updated_at, :null =>false
+      DateTime :deleted_at, :index => true
+      Integer :is_deleted, :null =>false
```
Comment (rakshasa) [inline vnet/db/migrations/0006_filters.rb]: This should be:

```
Integer :is_deleted, :null=>false, :default=>0
```

File: vnet/db/migrations/0006_filters.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+      Bignum :ipv4_src_address, :null => false
+      Bignum :ipv4_dst_address, :null => false
+      Integer :ipv4_src_prefix, :null => false
+      Integer :ipv4_dst_prefix, :null => false
+      Integer :port_src
+      Integer :port_dst
+      
+      String :protocol, :null => false
+      FalseClass :passthrough, :null => false
```
Comment (rakshasa) [inline vnet/db/migrations/0006_filters.rb]: Missing a 'unique' that includes filter_id and every column above except passthrough.

File: vnet/lib/vnet/models/filter.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    taggable 'fil'
+
+    plugin :paranoia_is_deleted
+
+    one_to_many :filter_statics
+
+    many_to_one :interface
+
+    plugin :association_dependencies,
+    # 0001_origin
```
Comment (rakshasa) [inline vnet/lib/vnet/models/filter.rb]: Comment is wrong, should be the filter migration file's name.

File: vnet/lib/vnet/models/filter_static.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+
+module Vnet::Models
+
+  # TODO: Refactor.
+
+  class FilterStatic < Base
+
```
Comment (rakshasa) [inline vnet/lib/vnet/models/filter_static.rb]: Missing paranoia module.

File: vnet/lib/vnet/models/filter_static.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      if protocol_included(["tcp", "udp"])
+        errors.add(:port_src, "port not in valid range") if !(0..0xffff).member?(self.port_src)
+        errors.add(:port_dst, "port not in valid range") if !(0..0xffff).member?(self.port_dst)
+      elsif protocol_included(["arp", "icmp", "all"])
+        errors.add(:port_src, "port needs to be nil") if !self.port_src.nil?
+        errors.add(:port_dst, "port needs to be nil") if !self.port_dst.nil?
+      elsif protocol_included(["arp", "all"])
+        errors.add(:ipv4_src_address, "ip address needs to be 0") if (self.ipv4_src_address > 0) # needs inspection
+        errors.add(:ipv4_dst_address, "ip address needs to be 0") if (self.ipv4_dst_address > 0) # needs inspection
+      end
```
Comment (rakshasa) [inline vnet/lib/vnet/models/filter_static.rb]: This does not create any errors if the protocol is unknown.

File: vnet/lib/vnet/models/interface.rb [hunk 6]
```diff
@@ -39,6 +40,7 @@ class Interface < Base
     active_interfaces: :destroy,
     datapath_networks: :destroy,
     datapath_route_links: :destroy,
+    filters: :destroy,
```
Comment (rakshasa) [inline vnet/lib/vnet/models/interface.rb]: Filters is not part of 0001_origin, so put it at the bottom with the appropriate migration filename.

File: vnet/lib/vnet/node_api/filter.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      private
+
+      def dispatch_created_item_events(model)
+        dispatch_event(FILTER_CREATED_ITEM, model.to_hash)
+      end
+
+      def dispatch_deleted_item_events(model)
+        dispatch_event(FILTER_DELETED_ITEM, id: model.id)
+
+        # 0001_origin
```
Comment (rakshasa) [inline vnet/lib/vnet/node_api/filter.rb]: Same, use filter migration filename.

File: vnet/spec/vnet/core/filters/helpers2.rb [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  flow_create(
+    table: params[:table],
+    goto_table: params[:goto_table],
+    priority: params[:priority],
+    match_interface: params[:interface_id],
+    cookie: params[:id] | COOKIE_TYPE_FILTER2,
+    match: params[:match]
+  )
+end
+
```
Comment (rakshasa) [inline vnet/spec/vnet/core/filters/helpers2.rb]: What is the reason for this file? Are all these methods used, and why aren't they part of static.rb?

**Comment by rakshasa** [comment]
+1

**Comment by Metallion** [comment]
I found a quite significant bug (https://github.com/axsh/openvnet/issues/431) in my review but since this is an optional feature, we can already merge it to development and fix that bug (hopefully) before releasing version 0.9.

Will merge tomorrow.

**Pull request merged**